File: old-projects/vector-defence-silverlight-port/src/Towers/SlowingTower.ts
import { Game } from "../Game";
import { Util } from "../Util";
import { SlowingIndicator } from "../Particles/SlowingIndicator";
import { Tower } from "./Tower";

export class SlowingTower extends Tower {
  private static readonly slowDownTo = 0.5;

  constructor(x: number, y: number, addToGame = true) {
    super(x, y, addToGame);
    this.cost = 30;
    this.range = 70;
    this.millisecondsBetweenFiring = 1000;
  }

  override render(context: CanvasRenderingContext2D): void {
    const pulse = (Math.sin(performance.now() / 500) + 1) / 2;
    const yellow = Math.round(255 * pulse);
    const gradient = context.createRadialGradient(this.x, this.y, 1, this.x, this.y, Tower.towerRadius);
    gradient.addColorStop(0, "#000");
    gradient.addColorStop(0.9, `rgb(${yellow},${yellow},0)`);
    context.save();
    context.fillStyle = gradient;
    context.strokeStyle = "#fff";
    context.lineWidth = 1.5;
    context.beginPath();
    context.arc(this.x, this.y, Tower.towerRadius, 0, Math.PI * 2);
    context.fill();
    context.stroke();
    context.restore();
  }

  protected override animate(): void {
    if (this.canFireNow) {
      const monstersToAffect = Game.currentFrameMonsters
        .filter((monster) => Util.isMonsterWithinDistance(monster, this.x, this.y, this.range))
        .slice(0, this.level + 2);
      for (const monster of monstersToAffect) {
        monster.slowDown(SlowingTower.slowDownTo);
        new SlowingIndicator(this.x, this.y, monster);
      }
      this.timeFiring();
    }
  }
}

File: old-projects/vector-defence-silverlight-port/original-csharp-source/Towers/SlowingTower.cs
using System;
using System.Linq;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using VectorDefenceSL.Particles;

namespace VectorDefenceSL.Towers
{
    public class SlowingTower : Tower
    {
        private const double SlowDownTo = .5;

        private readonly Storyboard storyboard = new Storyboard();
        private readonly GradientStop startGradient = new GradientStop { Color = Colors.Black, Offset = 0 };
        private readonly GradientStop endGradient = new GradientStop { Color = Colors.Yellow, Offset = .9 };
        private readonly ColorAnimationUsingKeyFrames animation = new ColorAnimationUsingKeyFrames { RepeatBehavior = RepeatBehavior.Forever };

        private readonly Ellipse torrent = new Ellipse
        {
            StrokeThickness = 1.5,
            Stroke = new SolidColorBrush(Colors.White),
            Width = TowerRadius * 2,
            Height = TowerRadius * 2
        };

        public SlowingTower(Button btn) : base(btn, TowerBuildingMode.SlowingTower)
        {
            torrent.Fill = new RadialGradientBrush(new GradientStopCollection { startGradient, endGradient });
            ConfigureAnimation();
            storyboard.Pause();
            btn.MouseEnter += delegate
            {
                if (Game.IsPlaying)
                {
                    storyboard.Resume();
                }
            };
            btn.MouseLeave += delegate
            {
                storyboard.Pause();
            };
        }

        public SlowingTower(double x, double y) : base(x, y)
        {
            torrent.Fill = new RadialGradientBrush(new GradientStopCollection { startGradient, endGradient });
            Cost = 30;
            Range = 70;
            MillisecondsBetweenFiring = 1000;
            torrent.MouseLeftButtonDown += HandleTowerSelection;
            ConfigureAnimation();
            Game.Paused += storyboard.Pause;
            Game.Resumed += storyboard.Resume;
        }

        public override sealed void Render()
        {
            torrent.SetCanvasPosition(X - TowerRadius, Y - TowerRadius);
            Container.Children.Add(torrent);
        }

        public override void RemoveTowerFromCanvas()
        {
            base.RemoveTowerFromCanvas();
            Container.Children.Remove(torrent);
            Game.Paused -= storyboard.Pause;
            Game.Resumed -= storyboard.Resume;
        }

        protected override void Animate()
        {
            if (CanFireNow)
            {
                var monstersCloseBy = Game.CurrentFrameMonsters.Where(m => m.IsWithinDistance(X, Y, Range));
                var monstersToAffect = monstersCloseBy.Take(Level + 2).ToArray();
                foreach (var m in monstersToAffect)
                {
                    m.SlowDown(SlowDownTo);
                    new SlowingIndicator(X, Y, m);
                }
                TimeFiring();
            }
        }

        private void ConfigureAnimation()
        {
            animation.KeyFrames.Add(new LinearColorKeyFrame { Value = Colors.Black, KeyTime = KeyTime.FromTimeSpan(TimeSpan.FromSeconds(.5)) });
            animation.KeyFrames.Add(new LinearColorKeyFrame { Value = Colors.Yellow, KeyTime = KeyTime.FromTimeSpan(TimeSpan.FromSeconds(1.0)) });
            Storyboard.SetTarget(animation, endGradient);
            Storyboard.SetTargetProperty(animation, new PropertyPath(GradientStop.ColorProperty));
            storyboard.Children.Add(animation);
            torrent.Resources.Add("s", storyboard);
            storyboard.Begin();
        }
    }
}
